### PR TITLE
Fix bugs with gr.update

### LINKED ISSUE #2154: 3.2's gr.update(visible=X) causes value to be reset
### Describe the bug

Every time I use gr.update(visible=X) on an element, its value becomes nothing. sliders get reset to zero, gr.Radio elements lose selection, ...

Here's how elements look in 3.2 after they are shown (they are invisible initially, and are shown when SD Upscale option is picked)  
firefox_Ws0SSiTDv8

Here's same with gradio downgraded to 3.1.5:
firefox_49CiIPMs9c

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

https://github.com/AUTOMATIC1111/stable-diffusion-webui

### Screenshot

see description

Also here is js error log for below
firefox_CFKKtTSH6f

### Logs

```shell
no errors in python

same unrelated errors in javascript for both versions
```

### System Info

```shell
3.2 gradio, Windows 10, Firefox and Chrome
```

### Severity

blocking upgrade to latest gradio version

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description

Fixes #2154 #2156

During #2044, I made it so that `delete_none` would keep `None` in the `value` key. This was to let users set the value of the component to `null` in the front-end which would reset the component value. The problem is that `None` is the default of the `value` key in the `update` component method, so the value is always reset, even when users are not trying to reset the value!!

The solution I went for was to distinguish the "don't update the value" and "set the value to null" use cases. For the former, we use an enum for the default value of the update method and for the latter, the user can return `None` as before.

The enum I chose is `_Keywords.NO_VALUE`. It's not exposed at the top-level gradio package and it's marked "private" by the leading underscore so it should not be autocompleted in IDES. So it should mean that the user did not pass in a value for the `"value"` key.

To test:

```python
import gradio as gr

with gr.Blocks() as demo:
    radio = gr.Radio(label="algorithm", choices=["a", "b", "c"], value="a")
    slider = gr.Slider(label="threshold", value=0, minimum=-1, maximum=1, visible=False, interactive=True)
    style = gr.Dropdown(label="style", choices=["modern", "classical", "cartoon"], value="modern", visible=False, interactive=True)

    def make_visible(algorithm):
        is_visible = algorithm == "c"
        return {slider: gr.update(visible=is_visible), style: gr.update(visible=is_visible)}
    radio.change(make_visible, inputs=radio, outputs=[slider, style])

demo.launch()
```
The slider value and drop down should be persisted when clicking different radio buttons.

I also added `blocks_update` to `all_demos` - slider value should be persisted as well.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gradio/test_data/blocks_configs.py [hunk 1]
```diff
@@ -59,7 +58,6 @@
             "id": 29,
             "type": "json",
             "props": {
-                "value": None,
```
Comment (freddyaboulton) [inline gradio/test_data/blocks_configs.py:62]: Backing out changes from #2044

**Comment by abidlabs** [comment]
Just wanted to confirm a few things:

1. This would be a breaking change, right, since users can no longer use `None` to reset a component. But I guess it would not too breaking since this was just introduced in #2044? 

2. Now to reset the value of a component, you pass in `gr.VOID` as the parameter to `gr.update(value=)`? an alternative might be `gr.RESET` if the purpose of the component is to reset a value.

3. Finally, I wanted to ask whether you can also return `gr.VOID` from a function directly, instead of within a `gr.update(value=)`. My preliminary experiments suggest that this doesn't work although I haven't tested thoroughly yet. I think we should have the same behavior if a user: (a) returns a value `x` from a function or (b) returns `gr.update(value=x)` from a function

**Comment by abidlabs** [comment]
Also cc @aliabid94 for his feedback on this API

**Comment by freddyaboulton** [comment]
@abidlabs Thanks for reviewing! Good questions!

1. Yea I guess this is a breaking change but #2044 was also technically a breaking change 🙈 
2. Yea I thought about `gr.RESET` but this doesn't reset to the original value, but rather it makes the value `null` in the front-end which was the goal of #1672 . We can also do `gr.NULL` ?
3.  Right now the logic only works for updates. We can do it so you can return `gr.VOID` from a function but that would mean modifying all the component post process functions to replace `VOID` with None.

**Comment by williamberrios** [comment]
Hi @freddyaboulton. It seems that when the output is only slider, the example doesn't work as expected

`import gradio as gr

with gr.Blocks() as demo:
    radio = gr.Radio(label="algorithm", choices=["a", "b", "c"], value="a")
    slider = gr.Slider(label="threshold", value=0, minimum=-1, maximum=1, visible=False, interactive=True)

    def make_visible(algorithm):
        is_visible = algorithm == "c"
        return {slider: gr.update(visible=is_visible)}
    radio.change(make_visible, inputs=radio, outputs=[slider])

demo.launch(share = True)`

**Comment by abidlabs** [comment]
Hey @freddyaboulton could we use sentinels to distinguish between the user passing in `None` for the `value` parameter versus the user leaving the `value` parameter empty? Here's how numpy does it: https://stackoverflow.com/a/60157942/5209347

If this works, it has several advantages: (1) it's not a breaking change (2) it does not introduce any key words / additional syntax that a Gradio user needs to know

**Comment by freddyaboulton** [comment]
Hi @williamberrios ! Thanks for the find. That's a separate issue (not introduced in this PR). I filed an issue for that here: https://github.com/gradio-app/gradio/issues/2191

**Comment by freddyaboulton** [comment]
@abidlabs Thank you for the suggestion about "sentinel values". I think it does work - I implemented it by just setting the default value of `value` in the `update` methods to be the Keyword enum I introduced in this PR. 

This should be good for another look!

**Comment by abidlabs** [comment]
Will take a look now!

File: gradio/components.py [hunk 2]
```diff
@@ -59,6 +60,11 @@ class DataframeData(TypedDict):
 set_documentation_group("component")
 
 
+class _Keywords(Enum):
+    # Set value in front-end to null
+    NO_VALUE = "NO_VALUE"
```
Comment (abidlabs) [inline gradio/components.py:65]:
```suggestion
class _Keywords(Enum):
    NO_VALUE = "NO_VALUE"  # Used as a sentinel to determine if nothing is provided as a argument for `value` in `Component.update()`
```

File: test/test_blocks.py [hunk 3]
```diff
@@ -266,5 +266,20 @@ def test_blocks_do_not_filter_none_values_from_updates(io_components):
     )
 
 
+def test_blocks_does_not_replace_keyword_literal():
+
+    with gr.Blocks() as demo:
+        text = gr.Textbox()
+        btn = gr.Button(value="Reset")
+        btn.click(
+            lambda: gr.update(value="VOID"),
```
Comment (abidlabs) [inline test/test_blocks.py:275]:
```suggestion
            lambda: gr.update(value="NO_VALUE"),
```

File: test/test_blocks.py [hunk 4]
```diff
@@ -266,5 +266,20 @@ def test_blocks_do_not_filter_none_values_from_updates(io_components):
...
+            inputs=[],
+            outputs=text,
+        )
+
+    output = demo.postprocess_data(0, gr.update(value="VOID"), state=None)
+    assert output[0]["value"] == "VOID"
```
Comment (abidlabs) [inline test/test_blocks.py:281]:
```suggestion
    output = demo.postprocess_data(0, gr.update(value="NO_VALUE"), state=None)
    assert output[0]["value"] == "NO_VALUE"
```

**Comment by abidlabs** [comment]
Some minor comments, but LGTM! Nice fix

**Review by abidlabs — APPROVED** [review]

**Pull request merged**